@xen-orchestra/backups/_cleanVm.mts

```diff
@@ -0,0 +1,736 @@
+        delete vhdChainsToMerge[vhd]
+        return chain
+      }
+
+      if (!unusedVhds.has(vhd)) {
+        return [vhd]
+      }
+
+      // no longer needs to be checked
+      toCheck.delete(vhd)
+
+      const child = vhdChildren[vhd]
+      if (child !== undefined) {
+        const chain = getUsedChildChainOrDelete(child)
+        if (chain !== undefined) {
+          chain.unshift(vhd)
+          return chain
+        }
+      }
+
+      // no warning because a VHD can be unused for perfectly good reasons,
+      // e.g. the corresponding backup (metadata file) has been deleted
+      if (remove) {
+        logInfo('deleting unused VHD', { path: vhd })
+        unusedVhdsDeletion.push(VhdAbstract.unlink(handler, vhd))
+      }
+    }
+
+    toCheck.forEach(vhd => {
+      vhdChainsToMerge[vhd] = getUsedChildChainOrDelete(vhd)
+    })
+
+    // merge interrupted VHDs
+    for (const parent of interruptedVhds.keys()) {
+      // before #6349 the chain wasn't in the mergeState
+      const { chain, statePath } = interruptedVhds.get(parent)
+      if (chain === undefined) {
+        vhdChainsToMerge[parent] = [parent, vhdChildren[parent]]
+      } else {
+        vhdChainsToMerge[parent] = chain.map(vhdPath => handlerPath.resolveFromFile(statePath, vhdPath))
+      }
+    }
+
+    Object.values(vhdChainsToMerge).forEach(chain => {
+      if (chain !== undefined) {
+        toMerge.push(chain)
+      }
+    })
+  }
+
+  const metadataWithMergedVhd = {}
+  const doMerge = async () => {
+    await asyncMap(toMerge, async chain => {
+      const { finalVhdSize } = await limitedMergeVhdChain(handler, chain, {
+        logInfo,
+        logWarn,
+        remove,
+        mergeBlockConcurrency,
+      })
+      const metadataPath: string = vhdsToJSons[chain[chain.length - 1]] // all the chain should have the same metada file
+      metadataWithMergedVhd[metadataPath] = (metadataWithMergedVhd[metadataPath] ?? 0) + finalVhdSize
+    })
+  }
+
+  await Promise.all([
+    ...unusedVhdsDeletion,
+    toMerge.length !== 0 && (merge ? Task.run({ name: 'merge' }, doMerge) : () => Promise.resolve()),
+    asyncMap(unusedXvas, path => {
+      logWarn('unused XVA', { path })
+      if (remove) {
+        logInfo('deleting unused XVA', { path })
+        return handler.unlink(path)
+      }
+    }),
+    asyncMap(xvaSums, path => {
+      // no need to handle checksums for XVAs deleted by the script, they will be handled by `unlink()`
+      if (!xvas.has(path.slice(0, -'.checksum'.length))) {
+        logInfo('unused XVA checksum', { path })
+        if (remove) {
+          logInfo('deleting unused XVA checksum', { path })
+          return handler.unlink(path)
+        }
+      }
+    }),
+  ])
+
+  // update size for delta metadata with merged VHD
+  // check for the other that the size is the same as the real file size
+  type OpaqueRef = `OpaqueRef:${string}`;
+  await asyncMap(jsons, async metadataPath => {
+    const metadata: {
```

Comment:
* please split type in smaller, reusable one
* please check how we can  reuse the xenapi types of @xen-orchestra/lite